### PR TITLE
Refactor the plugin, add docs and test coverage

### PR DESCRIPTION
**Pull request by stefanjudis**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,178 +1,302 @@
 # contentful-metalsmith
+[![Build Status](https://travis-ci.org/contentful-labs/contentful-metalsmith.svg?branch=master)](https://travis-ci.org/contentful-labs/contentful-metalsmith)
+[![Coverage Status](https://coveralls.io/repos/github/contentful-labs/contentful-metalsmith/badge.svg?branch=refactor)](https://coveralls.io/github/contentful-labs/contentful-metalsmith?branch=refactor)
 
-A Metalsmith's plugin to get content from [Contentful](http://www.contentful.com)
+A Metalsmith's plugin to generate files using content from [Contentful](http://www.contentful.com)
```
Comment (mariobodemann) [inline README.md]: Can you remove the `'` from the Metalsmit`'`s?

File: README.md [hunk 2]
```diff
@@ -1,178 +1,302 @@
...
-    </body>
-  </html>
-  ```
-
-5. Enjoy
+metalsmith.use(require('contentful-metalsmith')({ 'access_token' : 'YOUR_CONTENTFUL_ACCESS_TOKEN' }))
+```
 
-# Usage
+**Global parameters:**
```
Comment (mariobodemann) [inline README.md]: Hmm, I'm not a fan of `:` in headlines. Aren't they already introducing a section?

File: README.md [hunk 3]
```diff
@@ -1,178 +1,302 @@
...
+        </tr>
+      {{/each}}
+    </tbody>
+  </table>
+</body>
+</html>
+```
 
-```yaml
+For further instruction on how to distribute your data stored in Contentful check the (#usage)[usage section].
```
Comment (mariobodemann) [inline README.md]: The brackets for the link are opposite `[]()`

File: README.md (hunk 2 above)
Comment (stefanjudis) [inline README.md]: That's not a new headline but rather a bold section. But yeah `¯\_(ツ)_/¯`

File: README.md [hunk 4]
```diff
@@ -1,178 +1,302 @@
...
+  content_type: post        # id of the given content type - can refer to a hash
+  entry_template: post.html # template found in `layouts` folder
+layout: posts.html
+---
+Your post overview content here
+```
 
-* The template rendered for the source file.
-* And the template rendered for each individual entry.
+**This source file will:**
```
Comment (mariobodemann) [inline README.md]: See above for my distrust in `:`.

File: README.md (hunk 3 above)
Comment (stefanjudis) [inline README.md]: oh man... I'm always messing these up. :D

File: lib/processor.spec.js [hunk 5]
```diff
@@ -0,0 +1,203 @@
+import test from 'ava'
```
Comment (johanneswuerbach) [inline lib/processor.spec.js]: Just out of curiosity: Where is this spec on the same level as the file coming from? Never seen that before in js-land.

File: lib/processor.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  validator.validateFile(file)
+  validator.validateFileAndOptions(file, options)
+
+  const spaceId = file.contentful.space_id || options.space_id
+  const host = file.contentful.host || options.host
+  const query = util.getEntriesQuery(file.contentful)
+
+  const client = createContentfulClient(options.access_token, spaceId, host)
```
Comment (johanneswuerbach) [inline lib/processor.js]: I think a shared client for all files would be better or the contentful internal rate-limiting logic and should also simplify testing as proxyquire would no longer be required.

File: lib/processor.spec.js [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    host: 'global-baz',
+    space_id: 'global-foo'
+  }).then(fileMap => {
+    t.is(typeof fileMap['article/john-doe.html'], 'object')
+    t.is(fileMap['article/john-doe.html'].contents, '')
+    t.is(fileMap['article/john-doe.html'].data, entriesToBeReturned[0])
+    t.is(fileMap['article/john-doe.html'].id, entriesToBeReturned[0].sys.id)
+    t.is(fileMap['article/john-doe.html'].contentType, 'article')
+    t.is(fileMap['article/john-doe.html'].layout, 'template.html')
+    t.is(fileMap['article/john-doe.html']._fileName, 'article/john-doe.html')
```
Comment (johanneswuerbach) [inline lib/processor.spec.js]: Don't really know whats best practice in tap, but I generally prefer a deep-eq over asserting each property as deep-eq ensures that really all properties are covered.

File: lib/processor.spec.js (hunk 5 above)
Comment (stefanjudis) [inline lib/processor.spec.js]: Not sure. :D

File: lib/processor.spec.js [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    }
+  )
+
+  processor.processFile({
+    contentful: {}
+  }, {
+    access_token: 'global-bar',
+    host: 'global-baz',
+    space_id: 'global-foo'
+  }).then(_ => t.end())
```
Comment (johanneswuerbach) [inline lib/processor.spec.js]: Instead `test.cb` I would just return the Promise as this ensures the test is also finished when the promise is rejected.

File: lib/processor.spec.js [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+  }, {
+    access_token: 'global-bar',
+    host: 'global-baz',
+    space_id: 'global-foo'
+  }).then(fileMap => {
+    t.is(typeof fileMap['awesome-file.html'], 'object')
+    t.is(fileMap['awesome-file.html'].data, entryToBeReturned)
+
+    t.end()
+  }).catch(console.log)
```
Comment (johanneswuerbach) [inline lib/processor.spec.js]: Same ^^^

File: lib/processor.js (hunk 6 above)
Comment (johanneswuerbach) [inline lib/processor.js]: Yes, missed that. I would still build an `{ space_id: client }` as all the client back-off logic works only per instance and static site generation is usually triggering a bunch of requests in a really short timeframe.

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,22 @@
+sudo: false
+language: node_js
+cache:
+  directories:
+    - node_modules
+notifications:
+  email: false
+node_js:
+  - '6'
+  - '4'
```
Comment (Khaledgarbaya) [inline .travis.yml]: is there a reason we only test for node js v4 and v6 , what about v5 ?

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+before_install:
+  - npm i -g npm@^2.0.0
```
Comment (Khaledgarbaya) [inline .travis.yml]: shouldn't be npm@latest ?

Comment (stefanjudis) [inline .travis.yml]: That came with the semantic release `setup` command. what do you think, what should we do here now?

Comment (Khaledgarbaya) [inline .travis.yml]: ok let's leave it like that then for now

**Pull request merged**